cpp/src/arrow/flight/transport/ucx/ucx_client.cc

```diff
@@ -0,0 +1,730 @@
+      : UcxClientStream(impl, std::move(conn)) {
+    writes_done_ = true;
+  }
+
+  bool ReadData(internal::FlightData* data) override {
+    if (finished_) return false;
+
+    bool success = true;
+    io_status_ = ReadImpl(data).Value(&success);
+
+    if (!io_status_.ok() || !success) {
+      finished_ = true;
+    }
+    return success;
+  }
+
+ private:
+  ::arrow::Result<bool> ReadImpl(internal::FlightData* data) {
+    ARROW_ASSIGN_OR_RAISE(auto frame, driver_->ReadNextFrame());
+
+    if (frame->type == FrameType::kHeaders) {
+      // Trailers, stream is over
+      ARROW_ASSIGN_OR_RAISE(auto headers, HeadersFrame::Parse(std::move(frame->buffer)));
+      RETURN_NOT_OK(headers.GetStatus(&server_status_));
+      return false;
+    }
+
+    RETURN_NOT_OK(driver_->ExpectFrameType(*frame, FrameType::kPayloadHeader));
+    PayloadHeaderFrame payload_header(std::move(frame->buffer));
+    RETURN_NOT_OK(payload_header.ToFlightData(data));
+
+    // DoGet does not support metadata-only messages, so we can always
+    // assume we have an IPC payload
+    ARROW_ASSIGN_OR_RAISE(auto message, ipc::Message::Open(data->metadata, nullptr));
+
+    if (ipc::Message::HasBody(message->type())) {
+      ARROW_ASSIGN_OR_RAISE(frame, driver_->ReadNextFrame());
+      RETURN_NOT_OK(driver_->ExpectFrameType(*frame, FrameType::kPayloadBody));
+      data->body = std::move(frame->buffer);
+    }
+    return true;
+  }
+};
+
+class WriteClientStream : public UcxClientStream {
+ public:
+  WriteClientStream(UcxClientImpl* impl, ClientConnection conn)
+      : UcxClientStream(impl, std::move(conn)) {
+    std::thread t(&WriteClientStream::DriveWorker, this);
+    driver_thread_.swap(t);
+  }
+  arrow::Result<bool> WriteData(const FlightPayload& payload) override {
+    std::unique_lock<std::mutex> guard(driver_mutex_);
+    if (finished_ || writes_done_) return Status::Invalid("Already done writing");
+    outgoing_ = driver_->SendFlightPayload(payload);
+    working_cv_.notify_all();
+    received_cv_.wait(guard, [this] { return outgoing_.is_finished(); });
+
+    auto status = outgoing_.status();
+    outgoing_ = Future<>();
+    RETURN_NOT_OK(status);
+    return true;
+  }
+  Status WritesDone() override {
+    std::unique_lock<std::mutex> guard(driver_mutex_);
+    if (!writes_done_) {
+      ARROW_ASSIGN_OR_RAISE(auto headers, HeadersFrame::Make({}));
+      outgoing_ =
+          driver_->SendFrameAsync(FrameType::kHeaders, std::move(headers).GetBuffer());
+      working_cv_.notify_all();
+      received_cv_.wait(guard, [this] { return outgoing_.is_finished(); });
+
+      writes_done_ = true;
+      auto status = outgoing_.status();
+      outgoing_ = Future<>();
+      RETURN_NOT_OK(status);
+    }
+    return Status::OK();
+  }
+
+ protected:
+  void JoinThread() {
+    try {
+      driver_thread_.join();
+    } catch (const std::system_error&) {
+      // Ignore
+    }
+  }
+  void DriveWorker() {
```

Comment:
The reason why we spawn a thread is because the UCX worker is created without multithread support, but Flight's API allows concurrent/multithreaded read and write access, so we have to push all of the UCX work onto a single thread. I found that multithread support (which locks the worker) to have a lot of overhead (due to calling `ucp_worker_progress` a lot on multiple threads), though, it's possible good use of something like `ucp_worker_wait` or the epoll-based interface may avoid that, but I didn't have much luck in my experiments with either of those.

Reply:
Are the ucx callbacks (`AmSendCallback`, `HandleIncomingActiveMessage`) invoked inside `ucp_wroker_progress()`?